File: skills/bug-tracing/scripts/debug-logger.js
const { appendFileSync, mkdirSync, writeFileSync, existsSync } = require('fs');
const { join } = require('path');

const LOG_DIR = join(process.cwd(), 'tmp');
const LOG_PATH = join(LOG_DIR, 'debug.log');

try { mkdirSync(LOG_DIR, { recursive: true }); } catch {}

/**
 * Write a structured debug trace to tmp/debug.log.
 * @param {string} hypothesis - Which hypothesis this trace validates (H1, H2, H3...)
 * @param {string} location - Code location identifier (e.g. "processOrder:entry")
 * @param {*} [data] - Any runtime data to capture
 */
function debugTrace(hypothesis, location, data) {
  const entry = JSON.stringify({
    t: new Date().toISOString(),
    h: hypothesis,
    loc: location,
    data: data !== undefined ? data : null,
  }) + '\n';
  try { appendFileSync(LOG_PATH, entry); } catch {}
}

/** Clear the log file before a new reproduction session. */
function debugClear() {
  try { writeFileSync(LOG_PATH, ''); } catch {}
}

/** Check if the log file exists and has content. */
function debugHasLogs() {
  try { return existsSync(LOG_PATH); } catch { return false; }
}

module.exports = { debugTrace, debugClear, debugHasLogs };

File: skills/bug-tracing/scripts/debug_logger.py
import json
import os
from datetime import datetime

LOG_DIR = os.path.join(os.getcwd(), 'tmp')
LOG_PATH = os.path.join(LOG_DIR, 'debug.log')

os.makedirs(LOG_DIR, exist_ok=True)


def debug_trace(hypothesis: str, location: str, data=None):
    """Write a structured debug trace to tmp/debug.log."""
    entry = json.dumps({
        "t": datetime.now().isoformat(),
        "h": hypothesis,
        "loc": location,
        "data": data,
    }) + "\n"
    try:
        with open(LOG_PATH, "a") as f:
            f.write(entry)
    except Exception:
        pass


def debug_clear():
    """Clear the log file before a new reproduction session."""
    try:
        with open(LOG_PATH, "w") as f:
            f.write("")
    except Exception:
        pass


def debug_has_logs() -> bool:
    """Check if the log file exists and has content."""
    try:
        return os.path.exists(LOG_PATH) and os.path.getsize(LOG_PATH) > 0
    except Exception:
        return False
